File: application/external/src/main/kotlin/io/github/siyual_park/application/external/dto/request/CreateTokenRequest.kt
package io.github.siyual_park.application.external.dto.request

import com.fasterxml.jackson.annotation.JsonIgnore
import io.github.siyual_park.application.external.dto.GrantType
import io.swagger.annotations.ApiModelProperty
import javax.validation.constraints.AssertTrue

data class CreateTokenRequest(
    @ApiModelProperty(name = "grant_type")
    val grantType: GrantType,
    val username: String? = null,
    val password: String? = null,
    @ApiModelProperty(name = "refresh_token")
    val refreshToken: String? = null,
) {
    @JsonIgnore
    @AssertTrue
    fun isValid(): Boolean {
        return when (grantType) {
            GrantType.PASSWORD -> username != null && password != null
            GrantType.REFRESH_TOKEN -> refreshToken != null
        }
    }
}

File: k6-tests/src/request/create-token-request.ts
type CreateTokenFromClientCredentials = {
  grantType: 'client_credentials';
  clientId: string;
  clientSecret: string;
};

type CreateTokenFromRefreshToken = {
  grantType: 'refresh_token';
  clientId: string;
  clientSecret: string;
  refreshToken: string;
};

type CreateTokenFromPassword = {
  grantType: 'password';
  clientId: string;
  clientSecret: string;
  username: string;
  password: string;
};

type CreateTokenRequest = CreateTokenFromClientCredentials | CreateTokenFromRefreshToken | CreateTokenFromPassword;

export default CreateTokenRequest;
